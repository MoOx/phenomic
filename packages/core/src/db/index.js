import merge from "deep-assign";

const glue = "$$";
const nullSub = "__null__";
const emptyDatabase: PhenomicDBRegistry = {};
let database: PhenomicDBRegistry = emptyDatabase;

function sortBy(sort = "date") {
  return (a, b) => {
    a = a.data[sort];
    b = b.data[sort];
    if (!a && !b) return 0;
    if (!a) return -1;
    if (!b) return 1;
    if (b > a) return -1;
    if (a > b) return 1;
    return 0;
  };
}

function getSublevel(
  sub: null | string | Array<string>,
  filter: ?string,
  filterValue: ?string
) {
  if (!Array.isArray(sub)) {
    sub = [sub === null ? nullSub : sub];
  }
  let db = database[sub.join(glue)] || [];
  if (filter && filterValue && filter !== "default") {
    db = db.filter(item => {
      if (Array.isArray(item.data[filter])) {
        return item.data[filter].indexOf(filterValue) > -1;
      }
      return item.data[filter] === filterValue;
    });
  }
  return db;
}

function putToSublevel(
  subName: null | string | Array<string>,
  value: PhenomicDBEntry
) {
  if (!Array.isArray(subName)) {
    subName = [subName === null ? nullSub : subName];
  }
  const subname = subName.join(glue);
  const sub = database[subname] || [];
  database = {
    ...database,
    [subname]: [...sub.filter(item => item.id !== value.id), value].sort(
      (a, b) => (b.id > a.id ? -1 : 1)
    )
  };
}

function updateToSublevel(
  subName: null | string | Array<string>,
  value: PhenomicDBEntry
) {
  if (!Array.isArray(subName)) {
    subName = [subName === null ? nullSub : subName];
  }
  const subname = subName.join(glue);
  const sub = database[subname] || [];
  database = {
    ...database,
    [subname]: [
      ...sub.filter(item => item.id !== value.id),
      merge({}, sub.find(item => item.id === value.id) || {}, value)
    ].sort((a, b) => (b.id > a.id ? -1 : 1))
  };
}

async function getDataRelation(fieldName, ids) {
  let partial = null;
  try {
    if (Array.isArray(ids)) {
      partial = await Promise.all(ids.map(id => db.getPartial(fieldName, id)));
    } else {
      partial = await db.getPartial(fieldName, ids);
    }
    return partial;
  } catch (error) {
    return ids;
  }
}

async function getDataRelations(fields) {
  const ids = Object.keys(fields);
  const resolvedValues = await Promise.all(
    ids.map(id => getDataRelation(id, fields[id]))
  );
  return ids.reduce((resolvedFields, id, index) => {
    resolvedFields[id] = resolvedValues[index];
    return resolvedFields;
  }, {});
}

class NotFoundError extends Error {
  constructor(...args) {
    super(...args);
    this.name = "NotFoundError";
  }
}

const db = {
  destroy(): Promise<void> {
    return new Promise(resolve => {
      database = emptyDatabase;
      resolve();
    });
  },
  async put(
    sub: null | string | Array<string>,
    id: string,
    value: PhenomicDBEntryInput = { data: {}, partial: {} }
  ): Promise<void> {
    await putToSublevel(sub, {
      data: value.data,
      partial: value.partial,
      id
    });
  },
  async update(
    sub: null | string | Array<string>,
    id: string,
    value: PhenomicDBEntryInput = { data: {}, partial: {} }
  ): Promise<any> {
    await updateToSublevel(sub, {
      data: value.data,
      partial: value.partial,
      id
    });
  },
  async get(
    sub: null | string | Array<string>,
    id: string
  ): Promise<PhenomicDBEntryDetailed> {
    const item = getSublevel(sub).find(item => item.id === id);
    if (typeof item === "undefined") {
      throw new NotFoundError("ID not found in database");
    }
    const { body, ...metadata } = item.data;
    const relatedData = await getDataRelations(metadata);
    return {
      id: id,
      value: {
        ...relatedData,
        ...(body ? { body } : {})
      }
    };
  },
  async getPartial(sub: string | Array<string>, id: string): Promise<mixed> {
    const item = getSublevel(sub).find(item => item.id === id);
    if (!item) {
      return id;
    }
    const type = typeof item.partial;
    if (type === "string" || type === "number" || type === "boolean") {
      return item.partial;
    }
    return { id, ...item.partial };
  },

  getList(
    sub: string | Array<string>,
    config: LevelStreamConfig = {},
    filter: string = "default",
    filterValue: string = ""
  ): Promise<Array<any>> {
    return new Promise(resolve => {
      let collection = getSublevel(sub, filter, filterValue);
      collection.sort(sortBy());
      if (config.reverse) {
        collection = collection.concat().reverse();
      }
      if (config.gte) {
        const index = collection.findIndex(item => item.id === config.gte);
        collection = index > -1 ? collection.slice(index) : collection;
      } else if (config.gt) {
        const index = collection.findIndex(item => item.id === config.gt);
        collection = index > -1 ? collection.slice(index + 1) : collection;
      } else if (config.lte) {
        const index = collection.findIndex(item => item.id === config.lte);
        collection = index > -1 ? collection.slice(0, index + 1) : collection;
      } else if (config.lt) {
        const index = collection.findIndex(item => item.id === config.lt);
        collection = index > -1 ? collection.slice(0, index) : collection;
      }
      if (typeof config.limit === "number") {
        collection = collection.slice(
          0,
          Math.min(config.limit, collection.length)
        );
      }

      Promise.all(
        collection.map(item =>
          db.getPartial(sub, item.id).then(value => {
            const type = typeof value;
            if (
              type === "string" ||
              type === "number" ||
              type === "boolean" ||
              Array.isArray(value)
            ) {
              return {
                id: item.id,
                value
              };
            }
            return {
              ...value,
              id: item.id
            };
          })
        )
      ).then(resolve);
    });
  }
};

export default db;
