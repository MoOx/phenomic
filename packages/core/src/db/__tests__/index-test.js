import db from "..";

describe("db", () => {
  beforeEach(() => {
    // $FlowFixMe whatever
    jest.resetModuleRegistry();
  });

  it("should be able to put & get a value", () => {
    return db
      .put("collection-1", "bar", {
        data: { title: "bar" },
        partial: { title: "bar" }
      })
      .then(() => db.get("collection-1", "bar"))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get an item with relations", () => {
    i++;
    return Promise.all([
      db.put("authors" + i, "foo", {
        data: { name: "foo-name", bio: "foo-bio" },
        partial: { name: "foo-name" }
      }),
      db.put("bar" + i, "foo", {
        data: { ["authors" + i]: ["foo"] },
        partial: { ["authors" + i]: ["foo"] }
      })
    ])
      .then(() => db.get("bar" + i, "foo"))
      .then(value => expect(value).toMatchSnapshot());
  });

  let i = 0;

  it("should be able to get a list", () => {
    return Promise.all([
      db.put("test-list", "baz", { data: {}, partial: { title: "title-baz" } }),
      db.put("test-list", "foo", { data: {}, partial: { title: "title-foo" } })
    ])
      .then(() => db.getList("test-list", {}, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a reversed list", () => {
    return Promise.all([
      db.put("test-reversed-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-reversed-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() =>
        db.getList("test-reversed-list", { reverse: true }, "default", "")
      )
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a list after (starting) some value", () => {
    return Promise.all([
      db.put("test-gte-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-gte-list", "eoo", {
        data: {},
        partial: { title: "title-eoo" }
      }),
      db.put("test-gte-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() => db.getList("test-gte-list", { gte: "eoo" }, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a list after some value", () => {
    return Promise.all([
      db.put("test-gt-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-gt-list", "eoo", {
        data: {},
        partial: { title: "title-eoo" }
      }),
      db.put("test-gt-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() => db.getList("test-gt-list", { gt: "eoo" }, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a list before (starting) some value", () => {
    return Promise.all([
      db.put("test-lte-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-lte-list", "eoo", {
        data: {},
        partial: { title: "title-eoo" }
      }),
      db.put("test-lte-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() => db.getList("test-lte-list", { lte: "eoo" }, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a list before some value", () => {
    return Promise.all([
      db.put("test-lt-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-lt-list", "eoo", {
        data: {},
        partial: { title: "title-eoo" }
      }),
      db.put("test-lt-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() => db.getList("test-lt-list", { lt: "eoo" }, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should be able to get a list with a limited count", () => {
    return Promise.all([
      db.put("test-limit-list", "baz", {
        data: {},
        partial: { title: "title-baz" }
      }),
      db.put("test-limit-list", "foo", {
        data: {},
        partial: { title: "title-foo" }
      })
    ])
      .then(() => db.getList("test-limit-list", { limit: 1 }, "default", ""))
      .then(value => expect(value).toMatchSnapshot());
  });

  it("should throw when value isn't there", () => {
    return db
      .get("error", "baz")
      .catch(error => expect(error).toMatchSnapshot());
  });

  it("should be able to get a list with a specific query", () => {
    return Promise.all([
      db.put("test-query-list", "foo", {
        data: { title: "title-foo", tags: ["a", "b"] },
        partial: { title: "title-foo" }
      }),
      db.put("test-query-list", "bar", {
        data: { title: "title-bar", tags: ["a", "c"] },
        partial: { title: "title-bar" }
      }),
      db.put("test-query-list", "baz", {
        data: { title: "title-baz", tags: ["b", "c"] },
        partial: { title: "title-baz" }
      })
    ])
      .then(() => db.getList("test-query-list", {}, "tags", "a"))
      .then(value => expect(value).toMatchSnapshot());
  });
});
