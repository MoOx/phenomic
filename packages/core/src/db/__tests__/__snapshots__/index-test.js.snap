// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`db should be able to get a list 1`] = `
Array [
  Object {
    "id": "baz",
    "title": "title-baz",
  },
  Object {
    "id": "foo",
    "title": "title-foo",
  },
]
`;

exports[`db should be able to get a list after (starting) some value 1`] = `
Array [
  Object {
    "id": "eoo",
    "title": "title-eoo",
  },
  Object {
    "id": "foo",
    "title": "title-foo",
  },
]
`;

exports[`db should be able to get a list after some value 1`] = `
Array [
  Object {
    "id": "foo",
    "title": "title-foo",
  },
]
`;

exports[`db should be able to get a list before (starting) some value 1`] = `
Array [
  Object {
    "id": "baz",
    "title": "title-baz",
  },
  Object {
    "id": "eoo",
    "title": "title-eoo",
  },
]
`;

exports[`db should be able to get a list before some value 1`] = `
Array [
  Object {
    "id": "baz",
    "title": "title-baz",
  },
]
`;

exports[`db should be able to get a list with a limited count 1`] = `
Array [
  Object {
    "id": "baz",
    "title": "title-baz",
  },
]
`;

exports[`db should be able to get a list with a specific query 1`] = `
Array [
  Object {
    "id": "bar",
    "title": "title-bar",
  },
  Object {
    "id": "foo",
    "title": "title-foo",
  },
]
`;

exports[`db should be able to get a reversed list 1`] = `
Array [
  Object {
    "id": "foo",
    "title": "title-foo",
  },
  Object {
    "id": "baz",
    "title": "title-baz",
  },
]
`;

exports[`db should be able to get an item with relations 1`] = `
Object {
  "id": "foo",
  "value": Object {
    "authors1": Array [
      Object {
        "id": "foo",
        "name": "foo-name",
      },
    ],
  },
}
`;

exports[`db should be able to put & get a value 1`] = `
Object {
  "id": "bar",
  "value": Object {
    "title": "bar",
  },
}
`;

exports[`db should throw when value isn't there 1`] = `[NotFoundError: ID not found in database]`;
